### PR TITLE
Refactored AE tests to add new Enclave tests for testing AEv2

### PR DESCRIPTION
**Pull request by lilgreenbird**

### CODE REVIEW

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java [hunk 1]
```diff
@@ -598,8 +598,10 @@ static String getResource(String key) {
                     "An error occurred when matching VALUES list to table columns. Please verify SQL syntax."},
             {"R_invalidValuesList", "An error occurred when reading VALUES list. Please verify SQL syntax."},
             {"R_enclaveNotSupported", "The SQL Server instance does not support enclave based computations."},
-            {"R_enclaveNoAttestationProtocol",
-                    "The \"enclaveAttestationProtocol\" connection property must be specified with \"enclaveAttestationUrl\"."},
+            {"R_enclaveAEdisabled",
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java]: No need for 2 different error strings here. You can combine them into one string, something like `\"columnEncryptionSetting\",  \"enclaveAttestationUrl\", \"enclaveAttestationProtocol\" properties are required to enable Always Encrypted with secure enclaves.`

File: src/test/java/com/microsoft/sqlserver/jdbc/AETest.java [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+        String enclaveAttestationUrl = TestUtils.getConfiguredProperty("enclaveAttestationUrl");
+        connectionStringAE = TestUtils.addOrOverrideProperty(connectionStringAE, "enclaveAttestationUrl",
+                (null != enclaveAttestationUrl) ? enclaveAttestationUrl : "http://blah");
+
+        String enclaveAttestationProtocol = TestUtils.getConfiguredProperty("enclaveAttestationProtocol");
+        connectionStringAE = TestUtils.addOrOverrideProperty(connectionStringAE, "enclaveAttestationProtocol",
+                (null != enclaveAttestationProtocol) ? enclaveAttestationProtocol : AttestationProtocol.HGS.toString());
+
+        try (SQLServerConnection con = PrepUtil.getConnection(connectionStringAE)) {
+            isAEv2 = TestUtils.isAEv2(con);
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AETest.java]: Please verify the exception here. Use TestUtils.formatErrorMsg() to check `R_enclaveNotSupported`.

File: src/test/java/com/microsoft/sqlserver/jdbc/AETest.java [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+    private void verifyEnclaveEnabled(Connection con) throws SQLException {
+        try (Statement stmt = con.createStatement(); ResultSet rs = stmt.executeQuery(
+                "SELECT [name], [value], [value_in_use] FROM sys.configurations WHERE [name] = 'column encryption enclave type';")) {
+            while (rs.next()) {
+                assertEquals("1", rs.getString(2));
+            }
+        }
+    }
+
+    private void checkAEv2() {
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AETest.java]: This is not used anywhere.

File: src/test/java/com/microsoft/sqlserver/jdbc/AETest.java [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+    }
+
+    /**
+     * Tests invalid connection property combinations.
+     */
+    @Test
+    public void testInvalidProperties() {
+        org.junit.Assume.assumeTrue(isAEv2);
+
+        // enclaveAttestationUrl and enclaveAttestationProtocol without "columnEncryptionSetting"
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AETest.java]: Please create a copy of the original connection string to not modify it.

File: src/test/java/com/microsoft/sqlserver/jdbc/AETest.java [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+            List<String> paths = new ArrayList<String>();
+            paths.add(javaKeyPath);
+            String serverName = connection.activeConnectionProperties
+                    .getProperty(SQLServerDriverStringProperty.SERVER_NAME.toString()).toUpperCase();
+
+            trustedKeyPaths.put(serverName, paths);
+            SQLServerConnection.setColumnEncryptionTrustedMasterKeyPaths(trustedKeyPaths);
+
+            SQLServerSecurityUtility.verifyColumnMasterKeyMetadata(connection, "My_KEYSTORE", "UnTrustedKeyPath",
+                    serverName, true, null);
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AETest.java]: `fail(TestResource.getResource("R_expectedFailPassed"));` here?

File: src/test/java/com/microsoft/sqlserver/jdbc/AETest.java [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+    /*
+     * Test getEnclavePackage with null enclaveSession
+     */
+    @Test
+    public void testGetEnclavePackage() {
+        org.junit.Assume.assumeTrue(isAEv2);
+
+        SQLServerVSMEnclaveProvider provider = new SQLServerVSMEnclaveProvider();
+        try {
+            provider.getEnclavePackage(null, null);
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AETest.java]: What exactly are we **testing** here?

File: src/test/java/com/microsoft/sqlserver/jdbc/AETest.java [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+    }
+
+    /*
+     * Test null Attestation response
+     */
+    @Test
+    public void testNullAttestationResponse() throws SQLServerException {
+        org.junit.Assume.assumeTrue(isAEv2);
+
+        try {
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AETest.java]: @SuppressWarnings("unused")

File: src/test/java/com/microsoft/sqlserver/jdbc/AETest.java [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+    /*
+     * Test bad Attestation response
+     */
+    @Test
+    public void testBadAttestationResponse() throws SQLServerException {
+        org.junit.Assume.assumeTrue(isAEv2);
+
+        try {
+            byte[] responseBytes = new byte[36];
+            AttestationResponse resp = new AttestationResponse(responseBytes);
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AETest.java]: @SuppressWarnings("unused")

File: src/test/java/com/microsoft/sqlserver/jdbc/AETest.java [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+            assertTrue(e.getMessage().matches(TestUtils.formatErrorMsg("R_InvalidHealthCert")));
+        } catch (Exception e) {
+            fail(TestResource.getResource("R_unexpectedException") + e.getMessage());
+        }
+    }
+
+    /*
+     * Negative Test - AEv2 not supported
+     */
+    @Tag(Constants.xSQLv15)
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AETest.java]: Use xSQLv14, AEv2 is technically supported against 2k19.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java [hunk 10]
```diff
@@ -138,85 +153,39 @@
...
-                fail(e.getMessage());
-            }
-        }
-
-        cmkJks = Constants.CMK_NAME + "_JKS";
-        cekJks = Constants.CEK_NAME + "_JKS";
+        AETestConnectionString = connectionString + ";sendTimeAsDateTime=false" + ";columnEncryptionSetting=enabled";
 
-        readFromFile(Constants.JAVA_KEY_STORE_FILENAME, "Alias name");
+        if (null == applicationClientID || null == applicationKey || null == keyIDs
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java]: Are these properties also required for AEv1 tests?

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/EnclaveTest.java [hunk 11]
```diff
@@ -0,0 +1,630 @@
+/*
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/EnclaveTest.java]: I am confused about packages. AETest.java is in com.microsoft.sqlserver.jdbc package, but EnclaveTest.java is in com.microsoft.sqlserver.jdbc.AlwaysEncrypted package.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/EnclaveTest.java [hunk 12]
```diff
@@ -0,0 +1,630 @@
...
+@Tag(Constants.xAzureSQLDB)
+@Tag(Constants.reqExternalSetup)
+public class EnclaveTest extends JDBCEncryptionDecryptionTest {
+
+    private boolean nullable = false;
+
+    @BeforeAll
+    public static void setupEnclave() throws TestAbortedException, Exception {
+        boolean isAEv2 = false;
+        try (SQLServerConnection con = PrepUtil.getConnection(AETestConnectionString, AEInfo)) {
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/EnclaveTest.java]: This code can be moved to a common place, I see it is duplicated in AETest.java too.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/EnclaveTest.java [hunk 13]
```diff
@@ -0,0 +1,630 @@
...
+import com.microsoft.sqlserver.testframework.Constants;
+import com.microsoft.sqlserver.testframework.PrepUtil;
+
+
+/**
+ * Tests Enclave Decryption and encryption of values
+ *
+ */
+@RunWith(JUnitPlatform.class)
+@Tag(Constants.xSQLv12)
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/EnclaveTest.java]: Shouldnt this be @Tag(Constants.xSQLv14)?

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/EnclaveTest.java [hunk 14]
```diff
@@ -0,0 +1,630 @@
...
+     */
+    @Test
+    public void testCharSpecificSetterWindows() throws SQLException {
+        org.junit.Assume.assumeTrue(System.getProperty("os.name").startsWith("Windows"));
+
+        try (SQLServerConnection con = PrepUtil.getConnection(AETestConnectionString, AEInfo);
+                SQLServerStatement stmt = (SQLServerStatement) con.createStatement()) {
+            String[] values = createCharValues(nullable);
+
+            testChars(stmt, cekWin, charTable, values, TestCase.NORMAL, true);
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/EnclaveTest.java]: How do you handle Windows Keystore tests on Unix?

**Comment by peterbae** [comment]
delete .SQLServerColumnEncryptionAzureKeyVaultProvider.java.swo file

File: src/test/java/com/microsoft/sqlserver/jdbc/AETest.java (hunk 4 above)
Comment (lilgreenbird) [reply src/test/java/com/microsoft/sqlserver/jdbc/AETest.java]: this doesn't change it, it returns a copy

**Pull request merged**